Rename some type names that had "Type"  in them

thockin opened the pull request:
Just a bit of redundancy removal.  Old names left for compat.

/kind cleanup

```release-note
NONE
```
 /sig network

thockin committed:
Rename IPFamilyPolicyType => IPFamilyPolicy

thockin committed:
Rename ServiceExternalTrafficPolicyType => ServiceExternalTrafficPolicy

thockin committed:
Rename ServiceInternalTrafficPolicyType => ServiceInternalTrafficPolicy

thockin commented:
/retest

dcbw commented on pkg/apis/core/types.go:
```diff
@@ -3594,17 +3594,18 @@ const (
 // internal traffic
-type ServiceInternalTrafficPolicyType string
+type ServiceInternalTrafficPolicy string
+type ServiceInternalTrafficPolicyType = ServiceInternalTrafficPolicy
```
@thockin missing the `// for backwards compat` comment here? Or maybe we don't need it?

dcbw replied:
@thockin missing the // for backwards compat comment here? Or maybe we don't need it?

fedebongio commented:
/remove-sig api-machinery

dcbw commented:
/milestone v1.23

khenidak commented:
/lgtm

**Q**: do we have rules around dep compatibility? this will break people who are using these constants/enums via go-client, right?

thockin commented:
I think I have left all the old types as aliases - did I miss some?

thockin commented:
Updated and rebased, still some oddness

@dims @cheftako for legacy cloud provider

thockin commented:
@jiahuif @apelisse 

This PR renames some types and adds aliases for the old names.  It seems to break openapi enums.

The 1st commit renames `IPFamilyPolicyType` -> `IPFamilyPolicy`.  It is listed as `+enum` but I see swagger.json does not list it as an enum - just string.  I see this for fields of `IPFamily` type too.  Why is +enum not working?

The 2nd commit renames `ServiceExternalTrafficPolicyType` -> `ServiceExternalTrafficPolicy`.  I see swagger.json CHANGE it from an enum to a string but I have no idea why.

Can you offer insight?  Am I doing something wrong?

jiahuif commented:
There is no assumption about naming of enum types. The type is treated as an enum type iff the +enum marker presents.

dims commented:
/approve
/cc @liggitt

liggitt commented on api/openapi-spec/v3/api__v1_openapi.json:
```diff
@@ -29011,12 +29011,8 @@
           "externalTrafficPolicy": {
-            "description": "externalTrafficPolicy denotes if this Service desires to route external traffic to node-local or cluster-wide endpoints. \"Local\" preserves the client source IP and avoids a second hop for LoadBalancer and Nodeport type services, but risks potentially imbalanced traffic spreading. \"Cluster\" obscures the client source IP and may cause a second hop to another node, bu ...
-            "type": "string",
-            "enum": [
-              "Cluster",
-              "Local"
-            ]
```
is this correct? surprised to see an externally visible change here

liggitt commented on api/openapi-spec/swagger.json:
```diff
@@ -10200,11 +10200,7 @@
         },
         "externalTrafficPolicy": {
-          "description": "externalTrafficPolicy denotes if this Service desires to route external traffic to node-local or cluster-wide endpoints. \"Local\" preserves the client source IP and avoids a second hop for LoadBalancer and Nodeport type services, but risks potentially imbalanced traffic spreading. \"Cluster\" obscures the client source IP and may cause a second hop to another node, but  ...
-          "enum": [
-            "Cluster",
-            "Local"
-          ],
```
same question here... is the externally visible change correct?

jiahuif replied:
No. However, the problem was the constant aliases breaking enum OpenAPI generator. I am now working on a fix.

jiahuif replied:
The problem is with `k8s.io/gengo`, trying to fix.

jiahuif commented:
/hold 
/assign
Investigating the enum generation issue.

jiahuif commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4160,15 +4160,24 @@ const (
+)
+
+// for backwards compat
+type ServiceExternalTrafficPolicyType = ServiceExternalTrafficPolicy
```
```go
// for backwards compat
// +enum
type ServiceExternalTrafficPolicyType = ServiceExternalTrafficPolicy
```
To workaround a k/gengo bug.

thockin replied:
Why does `IPFamily` not show up in swagger.json (field name IPFamilies) as an enum?

thockin replied:
I did as you suggest but it doesn't really work:

It half works for 1 case, but duplicates the enum values in the list, and doesn't work at all for the other cases.

leilajal commented:
/triage accepted

spiffxp commented:
/milestone v1.24
We are past code freeze and this doesn't appear to be test-related to make it for test freeze.  If there is an exception to allow this through past code freeze, please re-milestone

thockin committed:
WIP: compat enums

Note that this works for 1 case, but duplicates the enum values, and
doesn't work at all for the other cases.

dims commented:
/close

thockin commented:
Obsoleted by https://github.com/kubernetes/kubernetes/pull/110990 and https://github.com/kubernetes/kubernetes/pull/110991
